core/src/Borrower.sol

```diff
@@ -70,26 +84,56 @@ contract Borrower is IUniswapV3MintCallback {
         packedSlot.owner = owner;
     }
 
-    // TODO liquidations
-    function liquidate() external {
+    function liquidate(ILiquidator callee, bytes calldata data) external {
         require(!packedSlot.isInCallback);
 
-        int24[] memory positions_ = positions.read();
-
-        (, int24 currentTick, , , , , ) = UNISWAP_POOL.slot0();
         (uint256 liabilities0, uint256 liabilities1) = _getLiabilities();
         Prices memory prices = _getPrices();
-        Assets memory assets = _getAssets(
-            positions_,
-            Uniswap.FeeComputationCache(
-                currentTick,
-                UNISWAP_POOL.feeGrowthGlobal0X128(),
-                UNISWAP_POOL.feeGrowthGlobal1X128()
-            ),
-            prices
-        );
+        Assets memory assets = _getAssetsAndWithdrawPositions(positions.read(), prices);
+
+        require(!BalanceSheet.isHealthy(liabilities0, liabilities1, assets, prices));
```

Comment:
We start by checking that the account's balance sheet is unhealthy, i.e. that it can actually be liquidated. In the process of summing up assets, we've also withdrawn all Uniswap positions (so that money is now sitting in this contract, rather than Uniswap)

Reply:
Wouldn't this withdraw the positions regardless of whether the balance sheet is healthy or not? If so, wouldn't that pose problems since bad actors could withdraw users' positions even if they are solvent? With that said, if there is something I am missing, please enlighten me.